tests/parallel/true_parallel.rs

```diff
@@ -66,13 +66,15 @@ fn true_parallel_same_keys() {
         }
     });
 
-    // Thread 2 will sync barrier *just* before calling `sum`. Doesn't
-    // guarantee the race we want but makes it highly likely.
+    // Thread 2 will wait until Thread 1 has entered sum and then --
```

Comment:
> From reading the function names and comments, I would assume thread1 signals 1 upon entry, and then waits for 2 to begin summing. 

Both of these things are true -- it does those **immediately upon entering** the query function, and before the query has really done anything yet. Is this the part that is confusing? It sounds like it will be signalling/waiting *before* entering the query?

Do you have a suggestion for a clearer name?

Reply:
Okay I've taken time to step through line by line and understand where I was confused. I'll post my explanation here.

My understanding of how an `InProgress` marker appears in `db` was incorrect. I was following the control flow with the thought that an `InProgress` marker could only be set here:
```
for ch in key.chars() {
    sum += db.input(ch);
}
```
What I followed from this, is that `key` would still be the individual letters, and _not_ "abc". Therefore, the only way that `thread2` would find an `InProgress` is if `thread1` was executing those three lines at the same time and they happened to check the _same_ letter at the _same_ time.

What finally clicked was that `key` is not an individual letter, but the string "abc". That key has therefore already been updated to `InProgress` when `thread1` entered the query, and `thread2` therefore blocks right away.

In summary, what I was confused by was how `thread1` became unblocked because `thread2` would not find `InProgress` for any specific letter. That would have led to `thread2` never advancing past `wait_for` and thus, the test not passing.